### DISCUSSION TITLE
In the devkit, are loading files larger than 1 mb possible?

### DISCUSSION THREAD

**Discussion by elkaamee326** [CosmosOS/Cosmos#1721] [Q&A]
In the userkit, when I tried to load a 2.63 mb file, it just displayed, "CPU exception: 0XD." In the dev kit, is this issue fixed? Also, is the exception a bitmap problem, File System, or is it from the Bitmap ctor? My file worked when the size was 640x480 (and that size was 900kb) but when I changed it to 720x1280, it showed that exception. BMP details: 24bpp BGR, 720x1280 and 2.63 mb file.  Is the larger than 1mb issue fixed in the dev kit? Thx.

This is the VM output:

![image](https://user-images.githubusercontent.com/69808082/114755675-ec610a00-9d27-11eb-868b-f142814fb2cc.png)

This was when I put a try and catch block.

**Comment by quajak** [comment]
The problem is a file system one and has not been fixed in the devkit. The problem does not appear directly at 1mb but also for files slightly smaller.

**Reply by elkaamee326** [reply]
What specific Cosmos C# Implementations are involved with this issue? Also, where is the main implementation of `Stream`? Because I noticed many issues with the Cosmos `Stream` when loading specific files. Also, why does it give an exception or load the file incorrectly at around 1mb? Thx for the help :)

**Reply by quajak** [reply]
To be honest I don't fully recall what the problem with files around this size is, it could be a memory problem or us having a problem in the file system when reading this large files. Its very difficult debugging such large operations and I haven't gotten around to it. 

Most of the Stream work is done in `FatStream`, and most errors will be in there or the other file system code. 

If you want to take a stab at solving some of them, please go ahead. I am happy to help/guide you and answer any questions you have.

**Reply by elkaamee326** [reply]
Could this be the problem? `long xMaxReadableBytes = mDirectoryEntry.mSize - mPosition;` Also, what is `DirectoryEntry.mSize` and what is it for?

**Reply by quajak** [reply]
A `DirectoryEntry ` doesnt necessarily mean that its a Directory, it can also still be a File on the disk. `mSize` is the size of the File/Directory and therefore limits how much can still be read. Therefore, I don' think  `long xMaxReadableBytes = mDirectoryEntry.mSize - mPosition` is the problem, since we limit how much more we can read by how there is in the file and how much we have read already.

**Reply by elkaamee326** [reply]
Could it be a memory problem? Because when I loaded a 0.98mb file, it loaded the `.bmp` fine, except at the top of the image it showed a bunch of random pixels. Could this be because it doesn't have a sufficient amount of RAM to store / draw those last few pixels? (The area of random pixels (length and width) was approximately an array of `30x640`)...

**Reply by elkaamee326** [reply]
Also, when I tried to load a 2.63mb file with the devkit, it displayed "CPU Exception x05" and not x0D. Do you know what each of these error codes mean?

**Reply by quajak** [reply]
> Could it be a memory problem? Because when I loaded a 0.98mb file, it loaded the `.bmp` fine, except at the top of the image it showed a bunch of random pixels. Could this be because it doesn't have a sufficient amount of RAM to store / draw those last few pixels? (The area of random pixels (length and width) was approximately an array of `30x640`)...

I would expect cosmos to not run into memory issues quiet that quickly but thats certainly possible. If you create a bitmap with just a height and width of a similar size and fill the color array with some values, can you display it correctly? The 2.63mb case most likely means the image did overflow into code, which means that random stuff happens and we crash and burn in some way.

**Comment by elkaamee326** [comment]
I am currently working on a project, which its GUI is based off of this: https://github.com/nifanfa/Cosmos-GUI-Sample

but apparently, I set the screen resolution to 1360x768 with a background image.

![image](https://user-images.githubusercontent.com/69808082/116936487-edee6580-ac35-11eb-958a-b4968a401300.png)

As you can see the fps is at 8 (when it is supposed to be at 30. This causes a big cursor lag) and for some reason when I load too many apps the VM crashes. 
Also, the background image was loaded by taking a `.bmp` file and converting it into a byte array and when I loaded it through a byte array it worked, but for some reason the cursor movement was very slow. I even tried increasing VRAM but it didn't really benefit as well as RAM too. What is the problem and can this be resolved? Also, I know what is the main cause of this problem (the 22.3 mb image byte array compiled into a `.dll` to allow the IDE to not crash.  When the `.dll` was compiled, it was only 3.56 mb) Is there anything I can do, or anything that is required? Thanks very much, any reply is appreciated :)

**Reply by elkaamee326** [reply]
Is there any way I can get rid of the Cursor lag and VM freeze?

**Reply by quajak** [reply]
The VM freeze is most likely caused by cosmos running out of memory, which you cant fix, except trying to minimize how many objects you allocate. Getting the GC working is a current priority and should hopefully be done soon.
The cursor lag is most likely caused by inefficiencies in the canvas. Since this is not a standard cosmos gui driver, I dont know but I suspect by inefficient copying. The best bet would be to either decrease the resolution or ensure that only less of the screen is updated each frame.

**Reply by elkaamee326** [reply]
Will the `GC` be implemented in the Memory Manager?

**Reply by elkaamee326** [reply]
Also, if the memory ran out then why if I try to increase ram or vram it isn't very useful and it doesn't resolve the cursor lag?

**Reply by quajak** [reply]
The Garbage Collector will extend the Memory Manager to automatically free memory again. Until that it done, using the new Memory Manager does not make sense.

**Reply by elkaamee326** [reply]
Once implemented, will the memory manager (Including GC) help stop the lag for this project and also, what did you mean by memory? Did you mean the memory that is assigned to the graphics card aka VRAM? Or did you mean the System RAM that cosmos uses for the CPU and the main hardware? Or the memory that is dedicated to Cosmos (code defined and not virtual machine defined if present). Lastly, will having a GC help with performance too when cleaning up garbage and anything that is created for everytime that the graphics card updates? Would this be a possible solution without changing the screen resolution. Thank you for all the help :)

**Reply by quajak** [reply]
The GC will not fix the lag you are experiencing but will stop the terminal freeze. The only way you will get rid of the freezes is by optimizing your graphics code.

**Reply by elkaamee326** [reply]
Okay, thanks :)
